gninasrc/lib/cnn_scorer.cpp

```diff
@@ -295,6 +301,151 @@ void CNNScorer::get_net_output(Dtype& score, Dtype& aff, Dtype& loss) {
+    );
+  }
+
+  CHECK_EQ(ligand_smtypes.size(), ligand_coords.size());
+}
+
+// Extracts receptor atoms and coordinates
+// Flex and inflex coordinates are taken from the model's movable atoms
+// Flex coordinates are stored at the beginning, then inflex, then fixed
+void CNNScorer::setReceptor(const model& m){
+
+  sz size = m.get_movable_atoms().size() + m.get_fixed_atoms().size();
+
+  if(receptor_smtypes.size() == 0){ // Do once at setup
+
+    // Preallocation (only on first call)
+    receptor_smtypes.reserve(size);
+
+    // Insert flexible residues movable atoms
+    auto cbegin = m.get_movable_atoms().cbegin();
+    auto cend = m.get_movable_atoms().cbegin() + m.ligands[0].node.begin;
+    for(auto it = cbegin; it != cend; ++it){
+      smt origt = it->sm; // Original smina type
+      receptor_smtypes.push_back(origt);
+    }
+
+    // Get number of movable atoms in flexible residues
+    num_flex_atoms = receptor_smtypes.size();
+
+    // Insert inflex atoms
+    cbegin = m.get_movable_atoms().cbegin() + m.m_num_movable_atoms;
+    cend = m.get_movable_atoms().cend();
+    for(auto it = cbegin; it != cend; ++it){
+      smt origt = it->sm; // Original smina type
+      receptor_smtypes.push_back(origt);
+    }
+
+    if(num_flex_atoms == 0){
+      // No atoms from flexible residues should be present
+      CHECK_EQ(receptor_coords.size(), 0);
+    }
+
+    // Insert fixed receptor atoms
+    cbegin = m.get_fixed_atoms().cbegin();
+    cend = m.get_fixed_atoms().cend();
+    for(auto it = cbegin; it != cend; ++it){
+      smt origt = it->sm; // Original smina type
+      receptor_smtypes.push_back(origt);
+    }
+  }
+
+  if(receptor_coords.size() == 0 ){ // Do once at setup
+
+    // Reserve memory, but size() == 0
+    receptor_coords.reserve(size);
+
+    // Append flex 
+    auto cbegin_flex = m.coordinates().cbegin();
+    auto cend_flex = m.coordinates().cbegin() + num_flex_atoms;
+    std::transform(cbegin_flex, cend_flex, std::back_inserter(receptor_coords),
+      [](const vec& coord) -> float3 {return float3({coord[0], coord[1], coord[2]});}
+    );
+
+    // Append inflex 
+    auto cbegin_inflex = m.coordinates().cbegin() + m.m_num_movable_atoms;
+    auto cend_inflex = m.coordinates().cend();
+    std::transform(cbegin_inflex, cend_inflex, std::back_inserter(receptor_coords),
+      [](const vec& coord) -> float3 {return float3({coord[0], coord[1], coord[2]});}
+    );
+
+    // Append rigid receptor
+    auto cbegin_rigid = m.get_fixed_atoms().cbegin();
+    auto cend_rigid = m.get_fixed_atoms().cend();
+    std::transform(cbegin_rigid, cend_rigid, std::back_inserter(receptor_coords),
+      [](const atom& a) -> float3 {
+        const vec& coord = a.coords;
+        return float3({coord[0], coord[1], coord[2]});
+      }
+    );
+    
+  }
+  else{ // Update flex coordinates at every call
+    auto cbegin = m.coordinates().cbegin();
+    auto cend = m.coordinates().cbegin() + num_flex_atoms;
```

Comment:
maybe a sanity CHECK here to make sure array is expected size?

Reply:
The total size is checked below (and in `CNNScorer::score`, where I moved some checks after removing `cnnopts.flexopt` and refactoring `setReceptor`), but I should add intermediate checks as well.